ethereum/eth/src/test/java/org/hyperledger/besu/ethereum/eth/manager/ethtaskutils/RetryingSwitchingPeerMessageTaskTest.java

```diff
@@ -0,0 +1,111 @@
+import org.hyperledger.besu.ethereum.eth.manager.EthProtocolManagerTestUtil;
+import org.hyperledger.besu.ethereum.eth.manager.RespondingEthPeer;
+import org.hyperledger.besu.ethereum.eth.manager.task.EthTask;
+
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+
+import org.junit.Test;
+
+/**
+ * Tests ethTasks that request data from the network, and retry until all of the data is received.
+ *
+ * @param <T> The type of data being requested from the network
+ */
+public abstract class RetryingSwitchingPeerMessageTaskTest<T> extends RetryingMessageTaskTest<T> {
+  protected Optional<EthPeer> responsivePeer = Optional.empty();
+
+  @Override
+  protected void assertResultMatchesExpectation(
+      final T requestedData, final T response, final EthPeer respondingPeer) {
+    assertThat(response).isEqualTo(requestedData);
+    responsivePeer.ifPresent(rp -> assertThat(rp).isEqualByComparingTo(respondingPeer));
+  }
+
+  @Test
+  public void completesWhenBestPeerEmptyAndSecondPeerIsResponsive()
+      throws ExecutionException, InterruptedException {
+    // Setup data to be requested and expected response
+
+    // Setup first unresponsive peer
+    final RespondingEthPeer firstPeer =
+        EthProtocolManagerTestUtil.createPeer(ethProtocolManager, 10);
+
+    // Setup first unresponsive peer
+    final RespondingEthPeer secondPeer =
+        EthProtocolManagerTestUtil.createPeer(ethProtocolManager, 9);
+
+    // Execute task and wait for response
+    final T requestedData = generateDataToBeRequested();
+    final EthTask<T> task = createTask(requestedData);
+    final CompletableFuture<T> future = task.run();
+
+    // First peer is not responsive
+    firstPeer.respond(RespondingEthPeer.emptyResponder());
+    // Second peer is responsive
+    secondPeer.respondTimes(
+        RespondingEthPeer.blockchainResponder(
+            blockchain, protocolContext.getWorldStateArchive(), transactionPool),
+        2);
+
+    responsivePeer = Optional.of(secondPeer.getEthPeer());
+
+    assertThat(future.isDone()).isTrue();
+    assertResultMatchesExpectation(requestedData, future.get(), secondPeer.getEthPeer());
+  }
+
+  @Test
+  public void completesWhenBestPeerTimeoutsAndSecondPeerIsResponsive()
+      throws ExecutionException, InterruptedException {
+    // Setup data to be requested and expected response
+    peerCountToTimeout.set(1);
+    // Setup first unresponsive peer
+    final RespondingEthPeer firstPeer =
+        EthProtocolManagerTestUtil.createPeer(ethProtocolManager, 10);
+
+    // Setup first unresponsive peer
+    final RespondingEthPeer secondPeer =
+        EthProtocolManagerTestUtil.createPeer(ethProtocolManager, 9);
+
+    // Execute task and wait for response
+    final T requestedData = generateDataToBeRequested();
+    final EthTask<T> task = createTask(requestedData);
+    final CompletableFuture<T> future = task.run();
+
+    // First peer timeouts
```

Comment:
I think the peerCountToTimeout.set(1) line could be moved here to make more sense?